### PR TITLE
Incorporate and Improve test coverage in the build/site #248

### PR DESCRIPTION
**Pull request by prb112**
- Add Examples to increase Test Coverage
	- parameters-contactdetail.[json/xml]
	- basic-type.[json/xml]
- Add coverage for get method tests
- Update fhir-parent to use the latest jacoco plugin
- Increase code coverage of fhir-config
- Update fhir-model to remove unused FHIRUtil method
- Add a switch to turn on/off codecov.io

Signed-off-by: Paul Bastide <pbastide@us.ibm.com>

### CODE REVIEW

File: fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+                }
+            }
+        }
+    }
+
+    public void runMethods(Object resource, Method[] methods) throws Exception {
+        for (Method method : methods) {
+            if (method.getName().startsWith("get")) {
+                method.invoke(resource);
+                assertTrue(true);
```
Comment (albertwang-ibm) [inline fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java]: Is this really useful?

File: fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+        CodeableConcept code = CodeableConcept.builder().id("1-2-3-4").text(string("Demo")).build();
+        Basic basic = Basic.builder().code(code).extension(extensions).build();
+
+        try (StringWriter writer = new StringWriter()) {
+            FHIRGenerator.generator(Format.XML).generate(basic, writer);
+
+            String out = writer.toString();
+            assertNotNull(out);
+            assertFalse(out.isEmpty());
+            System.out.println(out);
```
Comment (albertwang-ibm) [inline fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java]: do we really want to print them out always?

File: fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+            }
+        }
+
+        try (StringWriter writer = new StringWriter()) {
+            FHIRGenerator.generator(Format.JSON).generate(basic, writer);
+
+            String out = writer.toString();
+            assertNotNull(out);
+            assertFalse(out.isEmpty());
+            System.out.println(out);
```
Comment (albertwang-ibm) [inline fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java]: do we really want to print them out always?

File: fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+            try (ByteArrayInputStream in = new ByteArrayInputStream(out.getBytes())) {
+                Basic basicParsed = FHIRParser.parser(Format.JSON).parse(in).as(Basic.class);
+                assertNotNull(basicParsed);
+                assertTrue(true);
```
Comment (albertwang-ibm) [inline fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java]: not sure why assertTrue(true) which is always pass here ...

File: fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+
+            String out = writer.toString();
+            assertNotNull(out);
+            assertFalse(out.isEmpty());
+            System.out.println(out);
+
+            try (ByteArrayInputStream in = new ByteArrayInputStream(out.getBytes())) {
+                Basic basicParsed = FHIRParser.parser(Format.XML).parse(in).as(Basic.class);
+                assertNotNull(basicParsed);
+                assertTrue(true);
```
Comment (albertwang-ibm) [inline fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java]: does this always pass?

File: fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+                try (StringWriter writer = new StringWriter()) {
+                    FHIRGenerator.generator(Format.JSON).generate(resource, writer);
+                    String outJson = writer.toString();
+                    assertNotNull(outJson);
+                    assertFalse(outJson.isEmpty());
+                    
+                    try (ByteArrayInputStream in = new ByteArrayInputStream(outJson.getBytes())) {
+                        Resource resource2 = FHIRParser.parser(Format.JSON).parse(in).as(Resource.class);
+                        assertNotNull(resource2);
+                        assertTrue(true);
```
Comment (albertwang-ibm) [inline fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java]: always true?

File: fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+                try (StringWriter writer = new StringWriter()) {
+                    FHIRGenerator.generator(Format.XML).generate(resource, writer);
+                    String outXML = writer.toString();
+                    assertNotNull(outXML);
+                    assertFalse(outXML.isEmpty());
+                    
+                    try (ByteArrayInputStream in = new ByteArrayInputStream(outXML.getBytes())) {
+                        Resource resource2 = FHIRParser.parser(Format.XML).parse(in).as(Resource.class);
+                        assertNotNull(resource2);
+                        assertTrue(true);
```
Comment (albertwang-ibm) [inline fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java]: always true?

File: fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+    public void runMethods(Object resource, Method[] methods) throws Exception {
+        for (Method method : methods) {
+            if (method.getName().startsWith("get")) {
+                method.invoke(resource);
+                assertTrue(true);
+            }
+
+            if (method.getName().equals("hashCode")) {
+                method.invoke(resource);
+                assertTrue(true);
```
Comment (albertwang-ibm) [inline fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java]: always true?

File: fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+                if (resource instanceof Resource) {
+                    Resource r = (Resource) resource; 
+                    Resource r2 = r.toBuilder().build(); 
+                    assertTrue(r.equals(r2));
+                }
+            }
+
+            if (method.getName().equals("hasChildren")) {
+                method.invoke(resource);
+                assertTrue(true);
```
Comment (albertwang-ibm) [inline fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java]: always true?

Comment (prb112) [reply fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java]: Yes.  that's right.  it's an invocation, and asserts true as a pattern.

File: fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java (hunk 8 above)
Comment (prb112) [reply fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java]: Yes.  that's right.  it's an invocation, and asserts true as a pattern.

File: fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java (hunk 7 above)
Comment (prb112) [reply fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java]: Yes.  that's right.  it's an invocation, and asserts true as a pattern.

File: fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java (hunk 6 above)
Comment (prb112) [reply fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java]: Yes.  that's right.  it's an invocation, and asserts true as a pattern.

File: fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java (hunk 5 above)
Comment (prb112) [reply fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java]: Yes, I ran with ALL

File: fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java (hunk 3 above)
Comment (prb112) [reply fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java]: Yes.  that's right.  it's an invocation, and asserts true as a pattern.

File: fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java (hunk 4 above)
Comment (prb112) [reply fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java]: Yes.  that's right.  it's an invocation, and asserts true as a pattern.

File: fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java (hunk 2 above)
Comment (prb112) [reply fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java]: this only builds a test.  The corresponding content is in fhir-examples

File: fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java (hunk 1 above)
Comment (prb112) [reply fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java]: Yes, it's related to exercising the test framework.

File: fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java (hunk 4 above)
Comment (prb112) [reply fhir-model/src/test/java/com/ibm/fhir/model/test/ResourceCoverageTest.java]: These are used just to generate test data.

**Comment by prb112** [comment]
> LGTM, this definitely will increase the test coverage of all the resource types etc. even thought I'm not a fan of the assertTrue(true) pattern, but not a problem for me. BTW, maybe we can use the same mechanism to cover Type, inner classes also if higher coverage rate is pursued in the future.

Oh... I'm going to remove assertTrue's it's not a problem. I just couldn't switch branches. 
Hold please :)

**Comment by prb112** [comment]
removed assertTrue(true) per Albert's feedback

**Pull request merged**